crates/rpc/rpc/src/eth/gas_oracle.rs

```diff
@@ -0,0 +1,255 @@
+
+        // if we have stored a last price, then we check whether or not it was for the same head
+        if last_price.block_hash == header_hash {
+            return Ok(last_price.price)
+        }
+
+        // if all responses are empty, then we can return a maximum of 2*check_block blocks' worth
+        // of prices
+        //
+        // we only return more than check_block blocks' worth of prices if one or more return empty
+        // transactions
+        let mut current_hash = header_hash;
+        let mut results = Vec::new();
+        let mut populated_blocks = 0;
+
+        // we only check a maximum of 2 * max_block_history
+        for _ in 0..self.oracle_config.max_block_history * 2 {
+            let (parent_hash, block_values) = self
+                .get_block_values(current_hash, SAMPLE_NUMBER as usize)
+                .await?
+                .ok_or(EthApiError::UnknownBlockNumber)?;
+
+            if block_values.is_empty() {
+                results.push(U256::from(last_price.price));
+            } else {
+                results.extend(block_values);
+                populated_blocks += 1;
+            }
+
+            // break when we have enough populated blocks
+            if populated_blocks >= self.oracle_config.blocks {
+                break
+            }
+
+            current_hash = parent_hash;
+        }
+
+        // sort results then take the configured percentile result
+        let mut price = last_price.price;
+        if !results.is_empty() {
+            results.sort_unstable();
+            price = *results
+                .get((results.len() - 1) * self.oracle_config.percentile as usize / 100)
+                .expect("gas price index is a percent of nonzero array length, so a value always exists; qed");
+        }
+
+        // constrain to the max price
+        if let Some(max_price) = self.oracle_config.max_price {
+            if price > max_price {
+                price = max_price;
+            }
+        }
+
+        *last_price = GasPriceOracleResult { block_hash: header_hash, price };
+
+        Ok(price)
+    }
+
+    /// Get the `limit` lowest effective tip values for the given block. If the oracle has a
+    /// configured `ignore_price` threshold, then tip values under that threshold will be ignored
+    /// before returning a result.
+    ///
+    /// If the block cannot be found, then this will return `None`.
+    ///
+    /// This method also returns the parent hash for the given block.
+    async fn get_block_values(
+        &self,
+        block_hash: H256,
+        limit: usize,
+    ) -> EthResult<Option<(H256, Vec<U256>)>> {
+        // check the cache
+        let block = match self.cache.get_block(block_hash).await? {
+            Some(block) => block,
+            None => match self.client.block_by_hash(block_hash)? {
+                Some(block) => block,
+                None => return Ok(None),
+            },
```

Comment:
this branch is not needed, because if the cache is not able to find it (uses the same `block_by_hash` internally) then this won't either

Reply:
ah right, removed the branch